### PR TITLE
Support custom CSV termination.

### PR DESCRIPTION
**Pull request by heathdutton**
I am targeting this branch, because this is not BC breaking.

## Changelog
```markdown
### Added
- Added support for custom terminators with `CsvWriter` class
```

## Subject

A small feature addition to aid in the output of CSVs that are to be parsed by old/faulty programs/automation (mostly Microsoft). For example, if it's a program expecting "\r\n" for each line, it may only get the first line of the CSV and stop. Being able to control/manipulate the termination string (aka EOL) is pretty important for legacy support.

### CODE REVIEW

File: src/Writer/CsvWriterTerminate.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Exporter\Writer;
+
+/**
+ * Filter CSV output to replace the default terminator while supporting active streams.
+ */
+class CsvWriterTerminate extends \php_user_filter
```
Comment (greg0ire) [inline src/Writer/CsvWriterTerminate.php]: Please make this class final

File: src/Writer/CsvWriter.php [hunk 2]
```diff
@@ -58,21 +58,35 @@ class CsvWriter implements TypedWriterInterface
      */
     protected $withBom;
 
+    /**
+     * @var string
+     */
+    protected $terminate;
```
Comment (greg0ire) [inline src/Writer/CsvWriter.php]: Please make this property private (even though others are protected)

File: src/Writer/CsvWriter.php [hunk 3]
```diff
@@ -103,6 +117,10 @@ public function __construct($filename, $delimiter = ',', $enclosure = '"', $esca
     public function open()
     {
         $this->file = fopen($this->filename, 'w', false);
+        if ("\n" !== $this->terminate) {
+            stream_filter_register('filterTerminate', 'Exporter\Writer\CsvWriterTerminate');
```
Comment (kunicmarko20) [inline src/Writer/CsvWriter.php]: can you replace `'Exporter\Writer\CsvWriterTerminate'` with `CsvWriterTerminate::class`

**Comment by greg0ire** [comment]
> This branch cannot be rebased due to conflicts

Cannot merge, please rebase.

**Comment by heathdutton** [comment]
What conflicts? You approved these changes 7 days ago @greg0ire. I am confused.

**Comment by kunicmarko20** [comment]
@heathdutton there is something wrong with github, when we try to merge it says that there are conflicts, can you please try and do a rebase locally and if you get conflicts just resolve them (I had the same problems on my 2 PRs yesterday)

**Comment by heathdutton** [comment]
Just pushed a rebase, not sure if that's supposed to show up in this PR or not. Let me know if that didn't do the trick.

**Comment by OskarStark** [comment]
Everything looks good, but please add a testcase for `CsvWriterTerminate``

Thank you!

**Comment by heathdutton** [comment]
Added one. It was already being tested for use with CsvWriter, so I added a test for the filter in isolation. ¯\_(ツ)_/¯

**Comment by OskarStark** [comment]
Thank you 👌🏻👌🏻

File: src/Writer/CsvWriter.php [hunk 4]
```diff
@@ -58,27 +58,35 @@ class CsvWriter implements TypedWriterInterface
...
      */
     public function __construct(
         $filename,
         $delimiter = ',',
         $enclosure = '"',
         $escape = '\\',
         $showHeaders = true,
-        $withBom = false
+        $withBom = false,
+        $terminate = "\n"
```
Comment (kunicmarko20) [inline src/Writer/CsvWriter.php]: isn't this a BC Break?

Comment (kunicmarko20) [reply src/Writer/CsvWriter.php]: cc @sonata-project/contributors

Comment (greg0ire) [reply src/Writer/CsvWriter.php]: No, it isn't.

File: src/Writer/CsvWriterTerminate.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Exporter\Writer;
```
Comment (kunicmarko20) [inline src/Writer/CsvWriterTerminate.php]: This should have been `Sonata\Exporter\Writer`

Comment (kunicmarko20) [reply src/Writer/CsvWriterTerminate.php]: Nop, my bad, after checking stuff, I saw that on 1.x we still have `Exporter\....` and on master we use `Sonata\Exporter`...

Comment (greg0ire) [reply src/Writer/CsvWriterTerminate.php]: Careful with the automated merge then

**Pull request merged**